Remove QMSI sensor subsystem ADC driver and use Designware driver for ARC.

punitvara opened the pull request:
- Remove old adc_qmsi_ss driver and use DesignWare driver.
- Make DW driver to use dts.
- Update grove sensors ( light and temperature sensor) driver to use new ADC APIs.
- Update adc_api test for ARC board and it working fine. 
- Addressed all comments from #8857

anangl committed:
drivers: adc: Introduce reworked API

This commit replaces the API for ADC drivers with a reworked one.
As requested in the issue #3980, the adc_enable/adc_disable functions
are removed. Additionaly, some new features are introduced, like:
- asynchronous calls
- configuration of channels
- multi-channel sampling

Common parts of code that are supposed to appear in each implementation
of the driver (like locking, synchronization, triggering of consecutive
samplings) are provided in the "adc_context.h" file to keep consistency
with the SPI driver. Syscalls are no longer present in the API because
the functions starting read requests cannot use them, since they can be
provided with a callback that is executed in the ISR context, and there
is no point in supporting syscalls only for the channels configuration.

"adc_api" test is updated and extended with additional test cases,
with intention to show how the API is supposed to be used.
"adc_simple" test is removed as it does not seem to add much value.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: adc: Add shims for nrfx ADC and SAADC drivers

This commit adds translation layers to make nrfx drivers for the nRF
ADC (nRF51 series) and SAADC (nRF52 series) peripherals accessible via
the Zephyr's API. The SAADC peripheral is accessed using nrfx HAL only
as it turns out that usage of the nrfx driver in this case would be
inconvenient and would unnecessarily complicate the shim.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: nrf: Add ADC nodes and bindings for nRF SoCs

This commit adds ADC nodes to DTS files for nRF SoCs and introduces
corresponding  bindings for these nodes.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: nrf: Enable ADC nodes in DTS for nRF boards

ADC nodes are enabled in DTS for the following nRF development kits:
- nrf51_pca10028
- nrf52_pca10040
- nrf52_pca20020
- nrf52810_pca10040
- nrf52840_pca10056

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: adc_api: Add configurations for nRF5 boards

This adds to the adc_api test board-specific configurations
for the following boards:
- nrf51_pca10028
- nrf52_pca10040
- nrf52840_pca10056

"adc" is indicated as supported in the yaml files for these boards
so that they are included in sanitycheck.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: adc: Select HAS_DTS_ADC for all ADC drivers

To require relevant DTS entries on all platforms providing the ADC
driver.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

MaureenHelm committed:
adc: Convert mcux adc16 driver to the new adc api

Major rework of the mcux adc16 driver to convert it to the new adc api.
Currently supports a subset of the api features including synchronous
and asynchronous reads, and consecutive reads triggered by the kernel
timer.

Does not yet support some of the channel configuration options such as
gain and reference voltage because the hardware only allows these
options to be configured by peripheral instance rather than by channel.
The values are currently hardcoded in the driver, but in the future we
could introduce some flexibility per instance via device tree
attributes.

Does not yet support consecutive reads triggered by a hardware timer.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
tests: adc_api: Add support for kinetis boards

Adds board-specific macros to enable the adc_api test to run on kinetis
boards. This restores the test to support all the nxp boards that were
supported before the new adc api was introduced.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

punitvara committed:
drivers: grove: Modify light sensor

Replace old ADC apis with new ones.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: grove: Update temperature driver

Replace old ADC driver APIs with new ones.

Signed-off-by: Punit Vara <punit.vara@intel.com>

fallrisk committed:
sam: adc: Updated SAM ADC driver.

Updated to work with new ADC API.

Signed-off-by: Justin Watson <jwatson5@gmail.com>

punitvara committed:
drivers: Use Designware driver for sensor subsystem

Remove adc_qmsi_ss to use designware driver for sensor subsystem

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
dts: arc: Use dts tree for designware driver

Get required data from dts tree.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: adc: Update new ADC APIs

Replace new ADC apis for designware driver

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
tests: adc_api: Add ARC related parameters for ADC

Add ARC related parameters and modified test case to work
this test case on arc sensor subsystem.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: Kconfig.dw: Remove unnecessary Kconfig

Remove Kconfig which are replaced by dts and unnecessary macros which is
not useful anymore.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara commented:
@nashif  @MaureenHelm  Your review comments are addressed given on #8857. Please let me know if you feel I should provide any updates.

punitvara committed:
boards: quark_se_c1000_ss_devboard: Enable ADC node

Enable ADC node in dts.

Signed-off-by: Punit Vara <punit.vara@intel.com>

MaureenHelm reviewed (changes requested):
Thank you for fixing all the direct references to `adc_info_dev`, that part looks good now. There are additional comments in #9602 that still need to be addressed. Please also start running sanitycheck locally and fix the failures. You can do it like this:
```
$ sanitycheck -T tests/drivers/adc -T samples/sensor/grove_light -l
```

Please run gitlint on all of your commits, there is an issue with this one:
```
Commit b538745b9a:
1: UC3 Title does not follow [subsystem]: [subject]: "boards:quark_se_c1000_ss_devboard: Enable ADC node"
```

punitvara committed:
samples: grove_light: Provide device name from Kconfig

check return value for sample fetch. Conditional printk
added to print adc value.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
samples: grove_temperature: Use device name from Kconfig

Use Kconfig to get device name. Make LIBC conditional to get
output.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara commented:
@MaureenHelm I have addressed all your comments from #9602 and #8857 
Can you please review once ?

MaureenHelm reviewed (changes requested):
Please fix the CI failures. Instructions to run them locally are here: https://github.com/zephyrproject-rtos/zephyr/pull/9605#pullrequestreview-149307124

punitvara commented on drivers/adc/adc_dw.c:
```diff
@@ -153,17 +168,19 @@ static void adc_goto_normal_mode(struct device *dev)
 #endif
 
+/* TODO: Enable ADC power down capability */
+#ifdef CONFIG_ADC_DW_POWER_DOWN
```
ok. Removed.

punitvara commented on drivers/adc/adc_dw.c:
```diff
@@ -338,154 +455,90 @@ int adc_dw_init(struct device *dev)
-	};
+};
 
 static struct adc_config adc_config_dev = {
```
ok. Done

punitvara commented on samples/sensor/grove_temperature/prj.conf:
```diff
@@ -7,4 +7,5 @@ CONFIG_GROVE=y
 CONFIG_GROVE_TEMPERATURE_SENSOR_V1_X=y
-CONFIG_GROVE_LCD_RGB=y
+
+#CONFIG_GROVE_LCD_RGB=y
```
Done.

punitvara commented on drivers/adc/adc_dw.c:
```diff
@@ -200,116 +270,164 @@ static void adc_dw_enable(struct device *dev)
+	info->active_channels |= 1 << channel_id;
 
-	info->state = ADC_STATE_DISABLED;
+	adc_dw_enable(dev);
```
done.

punitvara commented on drivers/adc/adc_dw.c:
```diff
@@ -200,116 +270,164 @@ static void adc_dw_enable(struct device *dev)
+				(BIT(i) & info->active_channels)) {
+			info->channels |= BIT(i);
+		}
+	}
```
ok

punitvara commented on drivers/adc/adc_dw.c:
```diff
@@ -200,116 +270,164 @@ static void adc_dw_enable(struct device *dev)
-	sys_out32(ADC_INT_DSB|ADC_SEQ_PTR_RST, adc_base + ADC_CTRL);
+	/* There is no provision in the IP to set
+	 * gain and reference voltage settings. So
+	 * ignoring these settings.
```
ok

punitvara commented:
@MaureenHelm  Thanks for review. I addressed all your comments and also tested changes with grove sensors. Everything is working fine. Please have a look once.

punitvara commented on drivers/adc/Kconfig.dw:
```diff
@@ -77,20 +54,6 @@ config ADC_DW_FALLING_EDGE
 
 endchoice
 
-config ADC_DW_SAMPLE_WIDTH
```
Done

MaureenHelm commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -104,6 +104,14 @@
 #define ADC_1ST_CHANNEL_ID	0
 #define ADC_1ST_CHANNEL_INPUT	0
 
+#elif defined(CONFIG_ARC)
```
This still needs to be fixed

MaureenHelm commented on samples/sensor/grove_temperature/src/main.c:
```diff
@@ -41,9 +43,12 @@ void main(void)
-		sensor_sample_fetch(dev);
+		read = sensor_sample_fetch(dev);
+		if (read) {
+			printk("SAMLPE\n");
```
Typo

nashif replied:
please provide a meaningful message here, SAMPLE seems like a leftover debug message.

punitvara replied:
Done.

MaureenHelm commented on tests/drivers/adc/adc_api/testcase.yaml:
```diff
@@ -3,23 +3,4 @@ common:
-    extra_configs:
-      - CONFIG_ADC_QMSI_SAMPLE_WIDTH=8
-    platform_whitelist: quark_se_c1000_ss_devboard
+    filter: not CONFIG_SOC_QUARK_D2000
```
Instead of a filter, remove the adc line from `boards/x86/quark_d2000_crb/quark_d2000_crb.yaml`

punitvara replied:
Done

MaureenHelm reviewed (changes requested):
This is much better, thank you. Just a couple of minor issues left.

punitvara committed:
boards: quark_d2000_crb: Remove adc support

Remove ADC support from d2000 as it is not migrated to
new ADC apis.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
tests: adc_api: Remove sample width test scenario

Remove unnecessary sample width scenario.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara commented:
@nashif Done. Please let me know if anything need to be addressed.

nashif commented on samples/sensor/grove_temperature/prj.conf:
```diff
@@ -7,4 +7,3 @@ CONFIG_GROVE=y
 CONFIG_GROVE_TEMPERATURE_SENSOR_ADC_CHANNEL=10
 CONFIG_GROVE_TEMPERATURE_SENSOR=y
 CONFIG_GROVE_TEMPERATURE_SENSOR_V1_X=y
-CONFIG_GROVE_LCD_RGB=y
```
why is this being removed? isn't that part of the sample to display the sensor data on the LCD?

nashif replied:
On some boards you will need to use 2 pull-up resistors (10k Ohm) between the
SCL/SDA lines and 3.3V

punitvara replied:
@nashif After connecting two resistors of 10k connecting to SCA and SDA lines , I can see output temperature value on LCD. I am going to push it now.

MaureenHelm commented:
Looks good. I pulled your commits into #7691
